### DISCUSSION TITLE
Issues with ObjectIds and class-transformer

### DISCUSSION THREAD

**Discussion by smolinari** [TriPSs/nestjs-query#54] [General]
Hi @TriPSs,

As I was working on the Typegoose e2e tests, I noticed there were a ton of unit tests commented out. Digging deeper, the first problem I ran into was that the ObjectIds were getting reset by class-transformer and low and behold, it's a known issue and they don't want to fix it, They say the entities `_id` field needs a special `@Transform`, which is sort of silly to ask userland to add, as the `_id` field is standard in practically all Mongoose/ Mongodb entities. 

So, my question is, I've used a different transformer called @deepkit/type. If I can get it working to do the transformations would you be ok with that?

https://deepkit.io/library/type

Scott

**Comment by TriPSs** [comment]
I'm not quite know with deepkit, for the moment my prefference would be to keep using `class-validator` and also see what other think about this.

For the short term, does this impact the usage of the lib or can we just write a util that we can use in the e2e tests or even in the lib itself?

**Reply by smolinari** [reply]
It affects the lib for sure, but only in terms of the class-transformer-assembler and transforming objects to DTOs for finding related entities. This is done in the relation-query-services. 

What I can also attempt to build is a parallel assembler with `@deepkit/type` and only use that for the Mongoose and Typegoose transformations, where only they have the ObjectIds in use (I hope no one is using TypeORM with Mongodb...). 

Scott

**Reply by smolinari** [reply]
@TriPSs - I've come up with an interim and possibly final solution(?). It requires userland to add a decorator to any field with the return type of "ObjectId". The decorator is called: `@ObjectId` Haha! :grin: Btw, this is sort of what the class-transformer maintainers say is the "work around" for the problem. https://github.com/typestack/class-transformer/issues/494

For instance, for the `query-typegoose` unit tests, the `TestEntity`would be as follows:

```ts
export class TestEntity {
  @ObjectId()
  _id!: mongoose.Types.ObjectId

  id!: string

  @prop({ required: true })
  stringType!: string

  @prop({ required: true })
  boolType!: boolean

  @prop({ required: true })
  numberType!: number

  @prop({ required: true })
  dateType!: Date
  
  @ObjectId()
  @prop({ ref: TestReference, required: false })
  testReference?: Ref<TestReference> | string

  @ObjectId()
  @prop({ ref: TestReference, required: false })
  testReferences?: Ref<TestReference>[]
  
  @ObjectId()
  @prop({
    ref: 'TestReference',
    localField: '_id',
    foreignField: 'testEntity'
  })
  virtualTestReferences?: Ref<TestReference>[]
}
```
It would mean nothing else needs to be changed other than adding the decorator code, which `class-transformer` then uses to pass the proper value back from its `plainToClass` function. 

This is the decorator code: 

```ts
export const ObjectId = () => (target: unknown, propertyKey: string) => {
  Transform((params) => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access
    return params.obj[params.key]
  })(target, propertyKey)
}
```
Unfortunately I had to disable typescript warnings, because the `ObjectId` Type isn't available. 

What do you think?

Scott

**Reply by TriPSs** [reply]
Looks like a good solution :)

**Reply by smolinari** [reply]
Great, because it is already finished. :grin: I still have some e2e test issues (only 5!) and am having trouble troubleshooting. Once I get them fixed, both Mongoose and Typegoose packages will be complete.

Scott
